com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/app/discovery/ZigBeeNodeServiceDiscoverer.java

```diff
@@ -618,6 +624,11 @@ public void run() {
                         break;
                 }
 
+                if (closed) {
```

Comment:
I wonder if we should do this check at the very beginning of the `run` method to avoid the race condition where we start the discovery and at the same time remove the node. Then we won't call the `requestXYZ()` methods.

Reply:
The reason I put it after various methods is that is where the time is spent - ie this loop itself is very short from a time perspective (probably microseconds). However, the requests will take time (measured in tens of milliseconds, or even tens of seconds if the request times out). So, if we moved this check to the beginning, then it's fine if it gets caught, however, if it misses, and we start the requests, then we will end up rescheduling this.

Also, the shutdown method will stop the schedule, so in theory this should not run unless the run method was already executing when the shutdown was called. So, adding the check at the beginning of the method is probably not going to help at all.

Reply:
Good point, I agree that most time would be spend in the `requestXYZ()` calls.

Then the PR looks good to me.